### PR TITLE
*: asynchronous pd client, take 2

### PR DESCRIPTION
**Pull request by overvenus**
This is the second part of using an asynchronous PD client. It contains the following changes.

Done:
 - [x] Refactor trait `PdClient`.

Next PRs:
 - Add an AsyncWorker, it runs a tokio core. (Address https://github.com/pingcap/tikv/pull/1712#issuecomment-290088819)
 - Refactor PdWorker as AsyncWorker. (Address https://github.com/pingcap/tikv/pull/1712#issuecomment-290102625)
 - Add some metrics about asynchronous tasks.
 - Try to replace boxed `Future` with `impl Future` . (Address https://github.com/pingcap/tikv/pull/1712#issuecomment-290329925)

### CODE REVIEW

**Comment by overvenus** [comment]
PTAL @siddontang @BusyJay @disksing @andelf

**Comment by siddontang** [comment]
Great work, but please split your PR into two, one is for refactoring the request, the other is to change the client API.

File: src/pd/async/util.rs [hunk 1]
```diff
@@ -126,50 +127,52 @@ impl<Req, Resp, F> Request<Req, Resp, F>
...
             .boxed()
     }
 
+    fn receive(self) -> Result<Loop<Self, Self>> {
+        let ctx = self;
+        let done = ctx.reconnect_count == 0 || ctx.resp.is_some();
+        if done {
+            Ok(Loop::Break(ctx))
+        } else {
+            Ok(Loop::Continue(ctx))
```
Comment (siddontang) [inline src/pd/async/util.rs]: do we have any test to cover Continue here?

Comment (overvenus) [inline src/pd/async/util.rs]: Yes, `test_retry_async` https://github.com/pingcap/tikv/blob/master/tests/pd/test_rpc_client.rs#L117

File: src/pd/async/client.rs [hunk 2]
```diff
@@ -352,62 +361,73 @@ impl PdClient for RpcClient {
...
 
-    fn store_heartbeat(&self, stats: pdpb::StoreStats) -> Result<()> {
+    fn store_heartbeat(&self, stats: pdpb::StoreStats) -> PdFuture<()> {
         let mut req = pdpb::StoreHeartbeatRequest::new();
         req.set_header(self.header());
         req.set_stats(stats);
 
-        let resp = try!(do_request(self, |client| client.StoreHeartbeat(req.clone())));
-        try!(check_resp_header(resp.get_header()));
+        let request_factory = |client: &PDAsyncClient, req: pdpb::StoreHeartbeatRequest| {
```
Comment (siddontang) [inline src/pd/async/client.rs]: do we have any test to cover these futures?

**Comment by siddontang** [comment]
CI failed

**Comment by siddontang** [comment]
Rest LGTM

PTAL @BusyJay

File: src/pd/async/client.rs [hunk 3]
```diff
@@ -322,19 +322,28 @@ impl PdClient for RpcClient {
...
 
-    fn get_region_by_id(&self, region_id: u64) -> Result<Option<metapb::Region>> {
+    fn get_region_by_id(&self, region_id: u64) -> PdFuture<Option<metapb::Region>> {
         let mut req = pdpb::GetRegionByIDRequest::new();
         req.set_header(self.header());
         req.set_region_id(region_id);
 
-        let mut resp = try!(do_request(self, |client| client.GetRegionByID(req.clone())));
-        try!(check_resp_header(resp.get_header()));
+        let request_factory = |client: &PDAsyncClient, req: pdpb::GetRegionByIDRequest| {
```
Comment (BusyJay) [inline src/pd/async/client.rs]: It's not a request factory. A xxx_factory should produce xxx. `executor` makes more sense.

Comment (overvenus) [inline src/pd/async/client.rs]: No, it is a request that will be executed in the future.

Comment (BusyJay) [inline src/pd/async/client.rs]: What it returns is future rather than request. The request is req defined at L326.

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**